### Цільова функція

Нам необхідно максимізувати вартість вузлів, що будуються на основі мікросхем різних серій. Вартість вузлів для кожної серії мікросхем задана як:

\[ $
F = 50 \cdot x_1 + 48 \cdot x_2 + 45 \cdot x_3
$ \]

де:
- \( $ x_1 $ \) — кількість вузлів для мікросхеми K1 (АБО-НЕ),
- \( $ x_2 $ \) — кількість вузлів для мікросхеми K2 (І-НЕ),
- \( $ x_3 $ \) — кількість вузлів для мікросхеми K3 (⊕, І, 1).

### Обмеження

1. Для мікросхеми K1 (АБО-НЕ):
   \[ $
   2 \cdot x_1 + 3 \cdot x_2 + 4 \cdot x_3 \leq 20
   $ \]

2. Для мікросхеми K2 (І-НЕ):
   \[ $
   0 \cdot x_1 + 3 \cdot x_2 + 4 \cdot x_3 \leq 20
   $ \]

3. Для мікросхеми K3 (⊕, І, 1):
   \[ $
   8 \cdot x_1 + 10 \cdot x_2 + 6 \cdot x_3 \leq 64
   $ \]

### Умови для змінних

Змінні повинні задовольняти умови невід'ємності:
\[ $
x_1 \geq 0, \quad x_2 \geq 0, \quad x_3 \geq 0
$ \]


In [5]:
from scipy.optimize import linprog

# Цільова функція: максимізувати 50 * x1 + 48 * x2 + 45 * x3
# Оскільки linprog мінімізує функцію, множимо на -1
c = [-50, -48, -45]

# Обмеження:
# 2 * x1 + 3 * x2 + 4 * x3 <= 20
# 0 * x1 + 3 * x2 + 4 * x3 <= 20
# 8 * x1 + 10 * x2 + 6 * x3 <= 64
A = [
    [2, 3, 4],
    [0, 3, 4],
    [8, 10, 6]
]

# Праві частини обмежень
b = [20, 20, 64]

# Обмеження на змінні (x1, x2, x3 >= 0)
x_bounds = [(0, None), (0, None), (0, None)]

# Рішення задачі лінійного програмування
result = linprog(c, A_ub=A, b_ub=b, bounds=x_bounds, method='highs')

# Виведення результатів
if result.success:
    print("Рішення знайдено:")
    print(f"x1 = {result.x[0]:.2f}")
    print(f"x2 = {result.x[1]:.2f}")
    print(f"x3 = {result.x[2]:.2f}")
    print(f"Максимальна вартість F = {-result.fun:.2f}")
else:
    print("Не вдалося знайти розв'язок.")


Рішення знайдено:
x1 = 6.80
x2 = 0.00
x3 = 1.60
Максимальна вартість F = 412.00


### Перевірка отриманих значень:

Ми отримали оптимальні значення для змінних:

- \( $ x_1 = 6.80 $ \)
- \( $ x_2 = 0.00 $ \)
- \( $ x_3 = 1.60 $ \)

### Цільова функція:
Цільова функція має вигляд:
\[ $
F = 50 \cdot x_1 + 48 \cdot x_2 + 45 \cdot x_3
$ \]

#### 1. Перевірка значення цільової функції:
Підставляємо значення \( $ x_1 $ \), \( $ x_2 $ \) та \( $ x_3 $ \) у цільову функцію:
\[ $
F = 50 \cdot 6.80 + 48 \cdot 0.00 + 45 \cdot 1.60 = 340 + 0 + 72 = 412
$ \]
Максимальна вартість вузлів: **\( $ F = 412 $ \)**.

#### 2. Перевірка обмеження 1:
Обмеження:
\[ $
2 \cdot x_1 + 3 \cdot x_2 + 4 \cdot x_3 \leq 20
$ \]
Підставляємо значення:
\[ $
2 \cdot 6.80 + 3 \cdot 0.00 + 4 \cdot 1.60 = 13.60 + 0 + 6.40 = 20
$ \]
Оскільки \( $ 20 \leq 20 $ \), обмеження не порушується.

#### 3. Перевірка обмеження 2:
Обмеження:
\[ $
0 \cdot x_1 + 3 \cdot x_2 + 4 \cdot x_3 \leq 20
$ \]
Підставляємо значення:
\[ $
3 \cdot 0.00 + 4 \cdot 1.60 = 0 + 6.40 = 6.40
$ \]
Оскільки \( $ 6.40 \leq 20 $ \), обмеження не порушується.

#### 4. Перевірка обмеження 3:
Обмеження:
\[ $
8 \cdot x_1 + 10 \cdot x_2 + 6 \cdot x_3 \leq 64
$ \]
Підставляємо значення:
\[ $
8 \cdot 6.80 + 10 \cdot 0.00 + 6 \cdot 1.60 = 54.40 + 0 + 9.60 = 64
$ \]
Оскільки \( $ 64 \leq 64 $ \), обмеження не порушується.


# Контрольні питання

### 1. Умови існування та відсутності розв'язку задачі лінійного програмування

- **Умова існування розв'язку**:
  Задача лінійного програмування має розв'язок, якщо:
  1. Множина допустимих розв'язків неперервна (не порожня) та обмежена.
  2. Є точка, що задовольняє всі обмеження (включаючи умови невід'ємності змінних).
  
- **Умова відсутності розв'язку**:
  Задача не має розв'язку, якщо:
  1. Допустима множина порожня (немає жодного розв'язку, що задовольняє всі обмеження).
  2. Якщо задача має суперечливі обмеження або якщо неможливо досягти максимальної або мінімальної цільової функції (необмеженість).

---

### 2. Сутність симплекс-методу

Симплекс-метод — це метод для розв'язання задач лінійного програмування, який базується на переміщенні по гранях многогранника допустимих розв'язків від однієї вершини до іншої. 
- Починаючи з базового розв'язку, метод поступово покращує значення цільової функції, переміщуючись у напрямку, де вона покращується, доки не буде знайдений оптимальний розв'язок.
- Симплекс-метод працює шляхом пошуку найкращих (оптимальних) базисних розв'язків, застосовуючи правила оптимізації на кожному кроці.

---

### 3. Умова оптимальності знайденого плану симплекс-методом

Розв'язок задачі лінійного програмування вважається оптимальним, якщо:
- Для кожної невід'ємної змінної, що входить до базису, не існує покращення цільової функції за допомогою цієї змінної (це перевіряється через коефіцієнти в оптимальному базисі).
- Якщо всі коефіцієнти цільової функції, що відповідають змінним, які не входять в базис, не є від'ємними (для задачі на максимізацію). Це означає, що подальші зміни не покращують значення цільової функції.

---

### 4. Коли застосовується метод штучного базису?

Метод штучного базису застосовується:
- Коли задача лінійного програмування не має очевидного базисного розв'язку, тобто коли початковий базис не є допустимим.
- Для побудови допустимого початкового базису. Якщо задача не має допустимого початкового рішення, вводяться штучні змінні для створення початкового базису, і метод симплекс-прогресу використовується для подальшого пошуку оптимального розв'язку.
